allennlp/nlp_api.py

```diff
@@ -0,0 +1,261 @@
+            encoded_question = question_encoder(embedded_question)
+            # And we want to share this base context encoder across both the question and the
+            # passage, normally, so we "use question_encoder" as our fallback behavior.
+            passage_encoder = nlp_api.get_context_encoder("passage_encoder",
+                                                          fallback_behavior="use question_encoder")
+            encoded_passage = passage_encoder(embedded_passage)
+
+            # ... do bidirectional attention piece, merge passage and question representations
+            modeled_passage = merged_passage_representation
+            for i in range(self.num_modeling_layers):
+                # It would be odd to share stacked RNNs - these should be "new default" for a
+                # fallback behavior.  They will all be identical modules, with separate parameters,
+                # defined by the ``default_context_encoder`` function.  If you want them to have
+                # different structures (e.g., with more or less capacity as you get deeper), you
+                # can give them specific structures by passing in a mapping for each encoder name
+                # individually.
+                modeling_layer = nlp_api.get_context_encoder("modeling_layer_%d" % i,
+                                                             fallback_behavior="new default")
+                modeled_passage = modeling_layer(modeled_passage)
+
+            # ... predict span begin and span end, compute loss, return outputs
+
+    Then, when constructing the ``NlpApi`` object, you might give arguments that look roughly
+    like the following::
+
+        # Actual module instantiations here are made up, but this is the basic idea.
+        default_token_embedder = lambda: TokenEmbedder(pretrained_vectors=glove,
+                                                       character_dim=8)
+        default_context_encoder = lambda: BiLSTM(hidden_dim=256)
+        context_encoders = {'question_encoder': MyNewFancyRNN(hidden_dim=512)}
+
+        # These arguments can all be omitted, because they are defaults.
+        default_sentence_encoder = None
+        token_embedders = {}
+        sentence_encoders = {}
+
+    This would result in both the passage and the question using the same embedding layer
+    (instantiated from the ``default_token_embedder`` function), the question and passage sharing
+    an initial BiGRU encoding layer, and the modeling layers each having their own BiLSTM (each
+    instantiated from the ``default_context_encoder`` function).
+
+    Because you almost always just want to use a single embedding module for all of the TextFields
+    in your model, we give a default name for ``get_token_embedder``: ``"default"``.  This means
+    that you can simplify the code above a little bit::
+
+        def forward(self, question, passage, answer_span_begin=None, answer_span_end=None):
```

Comment:
Maybe I should have the simple example first, and show how to make it more flexible afterwards...